drivers/comparator/comparator_nrf_comp.c

```diff
@@ -0,0 +1,771 @@
+		return -EINVAL;
+	}
+
+	if (shim_nrf_comp_extrefsel_to_nrf(shim->extrefsel, &nrf->ext_ref)) {
+		return -EINVAL;
+	}
+
+	nrf->main_mode = NRF_COMP_MAIN_MODE_SE;
+
+	if (shim->th_down > 63 || shim->th_up > 63) {
+		return -EINVAL;
+	}
+
+	nrf->threshold.th_down = shim->th_down;
+	nrf->threshold.th_up = shim->th_up;
+
+	if (shim_nrf_comp_sp_mode_to_nrf(shim->sp_mode, &nrf->speed_mode)) {
+		return -EINVAL;
+	}
+
+	nrf->hyst = NRF_COMP_HYST_NO_HYST;
+
+#if NRF_COMP_HAS_ISOURCE
+	if (shim_nrf_comp_isource_to_nrf(shim->isource, &nrf->isource)) {
+		return -EINVAL;
+	}
+#else
+	if (shim->isource != COMP_NRF_COMP_ISOURCE_DISABLED) {
+		return -EINVAL;
+	}
+#endif
+
+	if (shim_nrf_comp_psel_to_nrf(shim->psel, &nrf->input)) {
+		return -EINVAL;
+	}
+
+	nrf->interrupt_priority = 0;
+	return 0;
+}
+
+static int shim_nrf_comp_diff_config_to_nrf(const struct comp_nrf_comp_diff_config *shim,
+					    nrfx_comp_config_t *nrf)
+{
+	nrf->reference = NRF_COMP_REF_AREF;
+
+	if (shim_nrf_comp_extrefsel_to_nrf(shim->extrefsel, &nrf->ext_ref)) {
+		return -EINVAL;
+	}
+
+	nrf->main_mode = NRF_COMP_MAIN_MODE_DIFF;
+	nrf->threshold.th_down = 0;
+	nrf->threshold.th_up = 0;
+
+	if (shim_nrf_comp_sp_mode_to_nrf(shim->sp_mode, &nrf->speed_mode)) {
+		return -EINVAL;
+	}
+
+#if NRF_COMP_HAS_HYST
+	if (shim->enable_hyst) {
+		nrf->hyst = NRF_COMP_HYST_ENABLED;
+	} else {
+		nrf->hyst = NRF_COMP_HYST_DISABLED;
+	}
+#else
+	if (shim->enable_hyst) {
+		return -EINVAL;
+	}
+#endif
+
+#if NRF_COMP_HAS_ISOURCE
+	if (shim_nrf_comp_isource_to_nrf(shim->isource, &nrf->isource)) {
+		return -EINVAL;
+	}
+#else
+	if (shim->isource != COMP_NRF_COMP_ISOURCE_DISABLED) {
+		return -EINVAL;
+	}
+#endif
+
+	if (shim_nrf_comp_psel_to_nrf(shim->psel, &nrf->input)) {
+		return -EINVAL;
+	}
+
+	nrf->interrupt_priority = 0;
+	return 0;
+}
+
+static int shim_nrf_comp_get_output(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return nrfx_comp_sample();
+}
+
+static int shim_nrf_comp_set_trigger(const struct device *dev,
+				     enum comparator_trigger trigger)
+{
+	shim_nrf_comp_stop();
+
+	switch (trigger) {
+	case COMPARATOR_TRIGGER_NONE:
+		shim_nrf_comp_data0.event_mask = 0;
+		break;
+
+	case COMPARATOR_TRIGGER_RISING_EDGE:
+		shim_nrf_comp_data0.event_mask = NRF_COMP_INT_UP_MASK;
+		break;
+
+	case COMPARATOR_TRIGGER_FALLING_EDGE:
+		shim_nrf_comp_data0.event_mask = NRF_COMP_INT_DOWN_MASK;
+		break;
+
+	case COMPARATOR_TRIGGER_BOTH_EDGES:
+		shim_nrf_comp_data0.event_mask = NRF_COMP_INT_CROSS_MASK;
+		break;
+	}
+
+	if (shim_nrf_comp_is_resumed()) {
+		shim_nrf_comp_start();
+	}
+
+	return 0;
+}
+
+static int shim_nrf_comp_set_trigger_callback(const struct device *dev,
+					      comparator_callback_t callback,
+					      void *user_data)
+{
+	shim_nrf_comp_stop();
+
+	shim_nrf_comp_data0.callback = callback;
+	shim_nrf_comp_data0.user_data = user_data;
+
+	if (callback != NULL && atomic_test_and_clear_bit(&shim_nrf_comp_data0.triggered, 0)) {
+		callback(dev, user_data);
+	}
```

Comment:
mmh the extra logic beyond `data->callback = cb` here seems like asking for inconsistent implementation between drivers, how about moving this to the api level in the `comparator.h` implementation?

Reply:
The issue here is that the underlying nrfx driver obfuscates the interrupt line, converting it to an event handler which this driver has to track. The expected behavior is that enabling the callback enables the interrupt, which, if pending/triggered, immediately triggers the irq, calling the callback. This is the case for the mcux_acmp driver/hardware, and indeed also for the nrf hardware. I would like to not have this rather dirty workaround for the nrf driver be part of the API :)